pylint/typing.py

```diff
@@ -113,7 +111,3 @@ class ManagedMessage(NamedTuple):
 
 AstCallback = Callable[["nodes.NodeNG"], None]
```

Comment:
Should this also be updated? Either way, I think it might be good to keep it in this file as we are likely to need to in other places.

Reply:
This also doesn't work, unfortunately. Just a short example
```py
from typing import Callable

from astroid import nodes
from pylint.extensions.typing import TypingChecker

AstCallback = Callable[[nodes.NodeNG], None]

var: list[AstCallback] = []

var.append(TypingChecker(None).visit_name)
```
At least pyright complains here since we would assign `Callable[[nodes.Name], None]` to `list[Callable[[nodes.NodeNG], None]]` which isn't possible.

Technically the parameter type for `AstCallback` would need to be a Union of all node subclasses. I don't think that is practical in this case though. Might be best to just use `Any` here.

Reply:
Hm, but normally if you type something to receive a parent class it should also be able to accept a subclass right? As in `BaseChecker` works for all methods/functions that accept any subclass of it. So why wouldn't that work here?

Reply:
Callable parameter are contravariant, thus the subclass relationship is inverted. A good explainer
https://blog.daftcode.pl/covariance-contravariance-and-invariance-the-ultimate-python-guide-8fabc0c24278

Reply:
> Ah, that explains it quite well. But wouldn't the same trick with a bounded `TypeVar` work?

TypeVars only really work if you use them at least twice. If not Pylance will even emit a warning
```
Type variable "..." has no meaning in this context
```

```py
from typing import Callable, TypeVar

class GrandParent: ...
class Parent(GrandParent): ...
class Child(Parent): ...

var: list[Callable[[Parent], None]] = []

def func_grandparent(x: GrandParent) -> None: ...
def func_parent(x: Parent) -> None: ...
def func_child(x: Child) -> None: ...

var.append(func_grandparent)
var.append(func_parent)

var.append(func_child)  # error


_ParentT = TypeVar("_ParentT", bound=Parent)
var2: list[Callable[[_ParentT], None]] = []
var2.append(func_child)  # error
```

Reply:
How do you invoke `pyright`. Is there a strictness flag I might be missing? I get:
```console
❯ pip install pyright
Requirement already satisfied: pyright in /Users/daniel/.pyenv/versions/3.10.0/envs/pylint-3.10.0/lib/python3.10/site-packages (1.1.244)
Requirement already satisfied: nodeenv>=1.6.0 in /Users/daniel/.pyenv/versions/3.10.0/envs/pylint-3.10.0/lib/python3.10/site-packages (from pyright) (1.6.0)
WARNING: You are using pip version 22.0.3; however, version 22.0.4 is available.
You should consider upgrading via the '/Users/daniel/.pyenv/versions/3.10.0/envs/pylint-3.10.0/bin/python3.10 -m pip install --upgrade pip' command.
❯ cat test.py
from typing import Callable

from astroid import nodes

from pylint.extensions.typing import TypingChecker
from pylint.lint import PyLinter

AstCallback = Callable[[nodes.NodeNG], None]

var: list[AstCallback] = []

var.append(TypingChecker(PyLinter()).visit_name)
❯ pyright test.py
No configuration file found.
No pyproject.toml file found.
stubPath /Users/daniel/DocumentenLaptop/Programming/Github/pylint/typings is not a valid directory.
Assuming Python platform Darwin
Searching for source files
Found 1 source file
0 errors, 0 warnings, 0 inform
```

Reply:
> How do you invoke `pyright`. Is there a strictness flag I might be missing?

`astroid` doesn't have a `py.typed` file yet. Thus all imports are basically ignored / set to `Any`. Try adding a `pyrightconfig.json` with
```json
{
    "typeCheckingMode": "strict",
    "useLibraryCodeForTypes": true,
}
```

If you use VSCode, you could also set `"python.analysis.typeCheckingMode": "strict"`.

```
test.py:3:6 - error: Stub file not found for "astroid" (reportMissingTypeStubs)
test.py:12:12 - error: Argument of type "(Name) -> None" cannot be assigned to parameter "__object" of type "AstCallback" in function "append"
    Type "(Name) -> None" cannot be assigned to type "AstCallback"
      Parameter 1: type "NodeNG" cannot be assigned to type "Name"
        "NodeNG" is incompatible with "Name" (reportGeneralTypeIssues)
```